File: app/src/main/kotlin/de/anubi1000/turnierverwaltung/strings/Strings.kt
package de.anubi1000.turnierverwaltung.strings

import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import de.anubi1000.turnierverwaltung.database.model.Participant

data class Strings(
    val appName: String = "Turnierverwaltung",

    val back: String = "Zurück",
    val cancel: String = "Abbrechen",
    val confirm: String = "Bestätigen",
    val create: String = "Neu",
    val delete: String = "Löschen",
    val edit: String = "Bearbeiten",
    val save: String = "Speichern",

    val amountOfResultsOnDisplay: String = "Anzahl Ergebnisse auf Anzeige",
    val basedOn: String = "Basiert auf",
    val dateOfTournament: String = "Datum des Turniers",
    val gender: String = "Geschlecht",
    val general: String = "Allgemein",
    val members: String = "Mitglieder",
    val name: String = "Name",
    val overview: String = "Übersicht",
    val participantsWithClub: String = "Teilnehmer mit diesem Verein",
    val search: String = "Suche",
    val searchTerm: String = "Suchbegriff",
    val startNumber: String = "Startnummer",
    val teamSize: String = "Teamgröße",

    val yesno: (Boolean) -> String = { arg1 ->
        if (arg1) "Ja" else "Nein"
    },

    val values: String = "Werte",
    val newValue: String = "Neuer Wert",
    val isAdded: String = "Wird addiert",
    val isAddedYesNo: (isAdded: Boolean) -> String = { arg1 ->
        buildString {
            append(isAdded)
            append(": ")
            append(yesno(arg1))
        }
    },

    val points: String = "Punkte",
    val inputPoints: String = "Punkte eintragen",
    val newRound: String = "Neue Runde",

    val genderSeparated: String = "Geschlechter getrennt",
    val genderName: (Participant.Gender) -> String = { arg1 ->
        when (arg1) {
            Participant.Gender.MALE -> "Männlich"
            Participant.Gender.FEMALE -> "Weiblich"
        }
    },

    val xDontExist: (name: String) -> String = { arg1 ->
        buildString {
            append("Es existieren noch keine ")
            append(arg1)
        }
    },
    val wantToDeleteX: (name: String) -> AnnotatedString = { arg1 ->
        buildAnnotatedString {
            append("Möchtest du ")
            withStyle(SpanStyle(fontWeight = FontWeight.Bold)) {
                append(arg1)
            }
            append(" wirklich löschen?")
        }
    },
    val editScreenTitle: (isEditMode: Boolean, name: String) -> String = { arg1, arg2 ->
        buildString {
            append(arg2)
            if (arg1) {
                append(" bearbeiten")
            } else {
                append(" erstellen")
            }
        }
    },

    val tournament: String = "Turnier",
    val tournaments: String = "Turniere",
    val noTournamentSelected: String = "Kein Turnier ausgewählt",
    val createTournament: String = "Turnier erstellen",
    val deleteTournament: String = "Turnier löschen",
    val editTournament: String = "Turnier bearbeiten",
    val openTournament: String = "Turnier öffnen",
    val wantToDeleteTournament: (name: String) -> AnnotatedString = { arg1 ->
        buildAnnotatedString {
            append("Möchtest du das Turnier ")
            withStyle(SpanStyle(fontWeight = FontWeight.Bold)) {
                append(arg1)
            }
            append(" wirklich löschen?")
        }
    },

    val participant: String = "Teilnehmer",
    val participants: String = "Teilnehmer",
    val noParticipantSelected: String = "Keinen Teilnehmer ausgewählt",
    val createParticipant: String = "Teilnehmer erstellen",
    val deleteParticipant: String = "Teilnehmer löschen",
    val editParticipant: String = "Teilnehmer bearbeiten",
    val wantToDeleteParticipant: (name: String) -> AnnotatedString = { arg1 ->
        buildAnnotatedString {
            append("Möchtest du den Teilnehmer ")
            withStyle(SpanStyle(fontWeight = FontWeight.Bold)) {
                append(arg1)
            }
            append(" wirklich löschen?")
        }
    },

    val team: String = "Team",
    val teams: String = "Teams",
    val noTeamSelected: String = "Kein Team ausgewählt",

    val club: String = "Verein",
    val clubs: String = "Vereine",
    val noClubSelected: String = "Kein Verein ausgewählt",
    val createClub: String = "Verein erstellen",
    val editClub: String = "Verein bearbeiten",

    val discipline: String = "Disziplin",
    val disciplines: String = "Disziplinen",
    val teamDiscipline: String = "Team-Disziplin",
    val teamDisciplines: String = "Team-Disziplinen",
    val newDiscipline: String = "Neue Disziplin",
    val noDisciplineSelected: String = "Keine Disziplin ausgewählt",
    val noDisciplineAvailable: String = "Keine Disziplinen vorhanden",
    val noTeamDisciplineAvailable: String = "Keine Team-Disziplinen vorhanden",

    val scoreboard: String = "Scoreboard",
    val showOnScoreboard: String = "Auf Scoreboard anzeigen",
)

File: Client/src/utils/strings.ts
import { Gender, type TeamScoreDisplayType } from "@/utils/api/schemas";

export const strings = {
  appName: "Turnierverwaltung",

  general: "Allgemein",
  other: "Sonstiges",
  results: "Ergebnisse",
  overview: "Übersicht",

  scoreboard: "Scoreboard",
  searchTerm: "Suchbegriff",
  newRound: "Neue Runde",
  round: "Runde",

  status: {
    loading: "Lädt...",
    listLoadingFailed: "Das Laden der Liste ist fehlgeschlagen",
    noEntriesAvailable: "Keine Einträge vorhanden",
    noResultsFound: "Keine Ergebnisse gefunden",
    overviewLoadingFailed: "Das Laden der Übersicht ist fehlgeschlagen",
    atLeastOneSelectedTable:
      "Es muss mindestens eine Tabelle ausgewählt werden",
  },

  actions: {
    login: "Anmelden",
    save: "Speichern",
    delete: "Löschen",
    cancel: "Abbrechen",
    edit: "Bearbeiten",
    back: "Zurück",
  },

  name: "Name",
  date: "Datum",
  teamSize: "Teamgröße",
  teamSizeFixed: "Teamgröße fix",
  memberCount: "Anzahl Mitglieder",
  startNumber: "Startnummer",
  gender: "Geschlecht",
  gendersSeparated: "Geschlechter getrennt",
  amountOfBestRoundsToShow: "Anzahl Runden auf Anzeige",
  isAdded: "Wird addiert",
  basedOn: "Basiert auf",
  type: "Typ",
  members: "Mitglieder",
  isShownInResult: "Wird in Ergebnis angezeigt",

  scores: {
    showOnScoreboard: "Auf Scoreboard anzeigen",
    downloadOverview: "Übersicht herunterladen",
    exportTables: "Tabellen zu exportieren",
    separateDocuments: "Separate Dokumente",
  },

  tournament: {
    item: "Turnier",
    items: "Turniere",
    create: "Neues Turnier",
    edit: "Turnier bearbeiten",
    selectOne: "Wähle ein Turnier aus",
    loadingError: "Beim Laden des Turniers ist ein Fehler aufgetreten",
    open: "Tunier öffnen",
    deleteDlg: {
      header: "Tunier löschen?",
      message: (name: string) =>
        `Möchtest du das Turnier "${name}" wirklich löschen?`,
    },
  },

  club: {
    item: "Verein",
    items: "Vereine",
    create: "Neuer Verein",
    edit: "Verein bearbeiten",
    selectOne: "Wähle einen Verein aus",
    loadingError: "Beim Laden des Vereins ist ein Fehler aufgetreten",
    count: "Anzahl Vereine",
    deleteDlg: {
      header: "Verein löschen?",
      message: (name: string) =>
        `Möchtest du den Verein "${name}" wirklich löschen?`,
    },
  },

  participant: {
    item: "Teilnehmer",
    items: "Teilnehmer",
    create: "Neuer Teilnehmer",
    edit: "Teilnehmer bearbeiten",
    selectOne: "Wähle einen Teilnehmer aus",
    loadingError: "Beim Laden des Teilnehmers ist ein Fehler aufgetreten",
    count: "Anzahl Teilnehmer",
    deleteDlg: {
      header: "Teilnehmer löschen?",
      message: (name: string) =>
        `Möchtest du den Teilnehmer "${name}" wirklich löschen?`,
    },

    editResults: "Ergebnisse bearbeiten",
  },

  team: {
    item: "Team",
    items: "Teams",
    create: "Neues Team",
    edit: "Team bearbeiten",
    selectOne: "Wähle ein Team aus",
    loadingError: "Beim Laden des Teams ist ein Fehler aufgetreten",
    count: "Anzahl Teams",
    notParticipatingInDisciplines:
      "Dieses Team nimmt an keiner Team-Disziplin teil",
    deleteDlg: {
      header: "Team löschen?",
      message: (name: string) =>
        `Möchtest du das Team "${name}" wirklich löschen?`,
    },
  },

  discipline: {
    item: "Disziplin",
    items: "Disziplinen",
    create: "Neue Disziplin",
    edit: "Disziplin bearbeiten",
    selectOne: "Wähle eine Disziplin aus",
    loadingError: "Beim Laden der Disziplin ist ein Fehler aufgetreten",
    count: "Anzahl Disziplinen",
    noAvailable: "Es sind Disziplinen vorhanden",
    deleteDlg: {
      header: "Disziplin löschen?",
      message: (name: string) =>
        `Möchtest du die Disziplin "${name}" wirklich löschen?`,
    },
  },

  teamDiscipline: {
    item: "Team-Disziplin",
    items: "Team-Disziplinen",
    create: "Neue Team-Disziplin",
    edit: "Team-Disziplin bearbeiten",
    selectOne: "Wähle eine Team-Disziplin aus",
    loadingError: "Beim Laden der Team-Disziplin ist ein Fehler aufgetreten",
    deleteDlg: {
      header: "Team-Disziplin löschen?",
      message: (name: string) =>
        `Möchtest du die Team-Disziplin "${name}" wirklich löschen?`,
    },
  },

  value: {
    items: "Werte",
    create: "Neuer Wert",
  },

  notFound: {
    pageNotFound: "Seite nicht gefunden",
    backToStart: "Zurück zum Start",
  },

  validation: {
    nameNeeded: "Es muss ein Name angegeben werden",
    numberNeeded: "Es muss eine Zahl angegeben werden",
    dateNeeded: "Es muss ein Datum angegeben werden",
    clubNeeded: "Es muss ein Verein angegeben werden",
    valueNeeded: "Es muss mindestens einen Wert geben",
    disciplineNeeded: "Es muss mindestens eine Disziplin angegeben werden",
    memberNeeded: "Es muss mindestens ein Mitglied geben",
    exactlyMembersNeeded: (teamSize: number) =>
      `Das Team muss genau ${teamSize} Mitglieder haben`,
  },

  formatting: {
    date: (date: number) => {
      const dateObj = new Date(date);

      return dateObj.toLocaleDateString("de-DE", {
        day: "2-digit",
        month: "long",
        year: "numeric",
      });
    },
    boolean: (bool: boolean) => {
      return bool ? "Ja" : "Nein";
    },
    gender: (gender: Gender) => {
      switch (gender) {
        case "Male":
          return "Männlich";
        case "Female":
          return "Weiblich";
      }
    },
    teamScoreDisplayType: (displayType: TeamScoreDisplayType) => {
      switch (displayType) {
        case "Normal":
          return "Normal";
        case "Nationcup":
          return "Nationencup";
        case "Triathlon":
          return "Triathlon";
      }
    },
  },
} as const;
